torsiondrive/qm_engine.py

```diff
@@ -42,11 +73,28 @@ def set_dihedral_constraints(self, dihedral_idx_values):
 
     def write_constraints_txt(self):
         """ write a constraints.txt file for geomeTRIC """
-        with open('constraints.txt', 'w') as outfile:
-            outfile.write("$set\n")
-            for d1, d2, d3, d4, v in self.dihedral_idx_values:
-                # geomeTRIC use atomic index starting from 1
-                outfile.write("dihedral %d %d %d %d %f\n" % (d1+1, d2+1, d3+1, d4+1, v))
+        if self.extra_constraints is None:
+            with open('constraints.txt', 'w') as outfile:
+                outfile.write("$set\n")
+                for d1, d2, d3, d4, v in self.dihedral_idx_values:
+                    # geomeTRIC use atomic index starting from 1
+                    outfile.write("dihedral %d %d %d %d %f\n" % (d1+1, d2+1, d3+1, d4+1, v))
+        else:
+            for key, value_list in self.extra_constraints.items():
+                if key == 'freeze' and len(value_list) > 0:
+                    constraints_string += '$' + key + '\n'
+                    for spec_tuple in value_list:
+                        constraints_string += ' '.join(spec_tuple) + '\n'
+                elif key == 'set':
+                    constraints_string += '$' + key + '\n'
+                    for spec_tuple in value_list:
+                        constraints_string += ' '.join(spec_tuple) + '\n'
+                    for d1, d2, d3, d4, v in self.dihedral_idx_values:
```

Comment:
It seems that if we have `extra_constraints` without "$set", the dihedral constraints generated by torsiondrive will not get written here.

Reply:
extra_constraints is initialized with `'set':[]` so there is always "something" there, even if it is empty.

Reply:
Oh I see. That's ok with me.